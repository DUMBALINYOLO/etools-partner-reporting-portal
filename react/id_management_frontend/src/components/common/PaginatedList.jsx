import React, {Component} from "react";
import {Paper, Typography} from "@material-ui/core";
import {
    Grid,
    Table,
    TableHeaderRow,
    TableRowDetail,
    PagingPanel,
    TableEditColumn,
    TableEditRow
} from "@devexpress/dx-react-grid-material-ui";
import {
    PagingState,
    CustomPaging, EditingState,
} from '@devexpress/dx-react-grid';
import {RowDetailState} from "@devexpress/dx-react-grid";
import {withStyles} from "@material-ui/core/styles";
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {connect} from 'react-redux';
import {expandedRowIds} from "../../actions";
import IconButton from "@material-ui/core/IconButton";
import Close from "@material-ui/icons/Close";
import Edit from "@material-ui/icons/Edit";
import Restore from "@material-ui/icons/SettingsBackupRestore";
import LoadingIndicator from "./LoadingIndicator";

const allowedPageSizes = [5, 10, 15];

const styleSheet = (theme) => {
    const paddingSmall = theme.spacing.unit * 2;
    const paddingBig = theme.spacing.unit * 3;

    return {
        container: {
            marginTop: paddingBig,
            position: 'relative'
        },
        loading: {
            backgroundColor: theme.palette.grey[200],
            pointerEvents: 'none'
        },
        header: {
            padding: `${paddingSmall}px 0 ${paddingSmall}px ${paddingBig}px`,
        }
    }
};

const DeleteButton = ({onClick}) => (
    <IconButton onClick={onClick} title="Delete row">
        <Close />
    </IconButton>
);

const EditButton = ({onClick}) => (
    <IconButton onClick={onClick} title="Edit row">
        <Edit/>
    </IconButton>
);

const RestoreButton = ({onClick}) => (
    <IconButton onClick={onClick} title="Restore row">
        <Restore/>
    </IconButton>
);

class PaginatedList extends Component {
    constructor(props) {
        super(props);

        this.onExpandedRowIdsChange = this.onExpandedRowIdsChange.bind(this);

        this.editCell = (showEdit, showDelete, showRestore) => (props) => {
            const {row} = props;
            const {onEdit, onDelete, onRestore} = this.props;

            return (
                <TableEditRow.Cell onValueChange={() => {
                }}>
                    {showEdit && <EditButton onClick={() => onEdit(row)}/>}
                    {showDelete && row.canBeDeleted && <DeleteButton onClick={() => onDelete(row)}/>}
                    {showRestore && row.canBeRestored && <RestoreButton onClick={() => onRestore(row)}/>}
                </TableEditRow.Cell>
            )
        }
    }

    header() {
        const {data: {count}, classes, page, pageSize} = this.props;

        const resultsFrom = (page - 1) * pageSize + 1;
        const resultsTo = Math.min(resultsFrom + pageSize - 1, count);

        return (
            <div className={classes.header}>
                <Typography variant="headline">
                    {`${resultsFrom}-${resultsTo} of ${count} results to show`}
                </Typography>
            </div>
        );
    }

    onExpandedRowIdsChange(ids) {
        const {dispatchExpandedRowIds} = this.props;

        dispatchExpandedRowIds(ids);

        if (this.props.onExpandedRowIdsChange) {
            this.props.onExpandedRowIdsChange(ids);
        }
    }

    render() {
        const {
            columns,
            data,
            expandedCell,
            classes,
            page,
            onPageChange,
            pageSize,
            onPageSizeChange,
            loading,
            expandedRowIds,
            showDelete,
            showEdit,
            showRestore
        } = this.props;

        const containerClasses = classNames(
            classes.container,
            {
                [classes.loading]: loading
            }
        );

        return (
            <Paper className={containerClasses}>
                <Grid rows={data.results} columns={columns}>
                    {this.header()}

                    <PagingState
                        currentPage={page - 1}
                        onCurrentPageChange={(page) => onPageChange(page + 1)}
                        pageSize={pageSize}
                        onPageSizeChange={onPageSizeChange}
                    />
                    <CustomPaging
                        totalCount={data.count}
                    />

                    {(showDelete || showEdit) &&
                    <EditingState
                        onCommitChanges={() => {
                        }}
                    />}

                    <Table/>
                    <TableHeaderRow/>
                    <RowDetailState expandedRowIds={expandedRowIds}
                                    onExpandedRowIdsChange={this.onExpandedRowIdsChange}/>

                    <TableRowDetail
                        contentComponent={({row}) => expandedCell(row)}
                    />

                    {(showDelete || showEdit || showRestore) &&
                    <TableEditColumn
                        width={64}
                        cellComponent={this.editCell(showEdit, showDelete, showRestore)}
                    />}

                    <PagingPanel pageSizes={allowedPageSizes}/>
                    {loading && <LoadingIndicator absolute/>}
                </Grid>
            </Paper>
        );
    }
}

PaginatedList.propTypes = {
    columns: PropTypes.array.isRequired,
    data: PropTypes.object.isRequired,
    expandedCell: PropTypes.func,
    page: PropTypes.number.isRequired,
    onPageChange: PropTypes.func.isRequired,
    onDelete: PropTypes.func,
    showDelete: PropTypes.bool,
    onEdit: PropTypes.func,
    showEdit: PropTypes.bool,
    onExpandedRowIdsChange: PropTypes.func,
};

const mapStateToProps = (state) => {
    const {expandedRowIds} = state;

    return {
        expandedRowIds
    };
};

const mapDispatchToProps = dispatch => {
    return {
        dispatchExpandedRowIds: ids => dispatch(expandedRowIds(ids))
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(withStyles(styleSheet)(PaginatedList));
