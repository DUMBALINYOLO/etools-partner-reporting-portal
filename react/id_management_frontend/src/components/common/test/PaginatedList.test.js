import React from 'react';
import {shallow, mockStore} from 'enzyme';
import toJSON from 'enzyme-to-json';
import {PaginatedList, mapStateToProps} from '../PaginatedList';

describe('PaginatedList component', () => {
    const columns = [];
    const columnExtensions = [];
    const alternativeSorting = [];
    const data = {};
    const expandedCell = jest.fn();
    const page = 451;
    const onPageChange = jest.fn();
    const onDelete = jest.fn();
    const showDelete = false;
    const showRestore = false;
    const allowSorting = false;
    const loading = false;
    const onEdit = jest.fn();
    const showEdit = false;
    const onExpandedRowIdsChange = jest.fn();
    const classes = {};

    const wrapper = shallow(<PaginatedList
        columns={columns}
        columnExtensions={columnExtensions}
        data={data}
        expandedCell={expandedCell}
        page={page}
        onPageChange={onPageChange}
        onDelete={onDelete}
        showDelete={showDelete}
        showRestore={showRestore}
        allowSorting={allowSorting}
        loading={loading}
        onEdit={onEdit}
        showEdit={showEdit}
        onExpandedRowIdsChange={onExpandedRowIdsChange}
        classes={classes}
    />);

    it('renders the component', () => {
        expect(wrapper.dive().length).toBe(1);
        expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it('renders the header', () => {
        const instance = wrapper.instance();
        const header = shallow(instance.header());

        expect(header.find('div')).toBeTruthy();
    });

    it('runs mapStateToProps correctly', () => {
        const state = {expandedRowIds: 'test'};
        const map = mapStateToProps(state);

        expect(map).toEqual(state);
    });
});