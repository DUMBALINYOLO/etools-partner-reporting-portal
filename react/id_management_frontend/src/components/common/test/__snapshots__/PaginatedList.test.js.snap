// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginatedList component renders the component 1`] = `
<WithStyles(Paper)
  className=""
>
  <Grid$$1
    columns={Array []}
  >
    <div>
      <WithStyles(Typography)
        variant="headline"
      >
        NaN-NaN of undefined results to show
      </WithStyles(Typography)>
    </div>
    <SortingState
      columnExtensions={Array []}
      columnSortingEnabled={true}
      defaultSorting={Array []}
      onSortingChange={[Function]}
    />
    <PagingState
      currentPage={450}
      defaultCurrentPage={0}
      defaultPageSize={10}
      onCurrentPageChange={[Function]}
    />
    <CustomPaging
      totalCount={0}
    />
    <Table$$1
      messages={Object {}}
      rowComponent={[Function]}
    />
    <TableHeaderRow$$1
      messages={Object {}}
      showSortingControls={false}
    />
    <RowDetailState
      defaultExpandedRowIds={Array []}
      onExpandedRowIdsChange={[Function]}
    />
    <TableRowDetail$$1
      contentComponent={[Function]}
    />
    <PagingPanel$$1
      messages={Object {}}
      pageSizes={
        Array [
          5,
          10,
          15,
        ]
      }
    />
  </Grid$$1>
</WithStyles(Paper)>
`;
